src/cpp/transport/serial/MultiSerialAgentLinux.cpp

```diff
@@ -0,0 +1,224 @@
+// limitations under the License.
+
+#include <uxr/agent/transport/serial/MultiSerialAgentLinux.hpp>
+#include <uxr/agent/utils/Conversion.hpp>
+#include <uxr/agent/logger/Logger.hpp>
+
+#include <unistd.h>
+
+namespace eprosima {
+namespace uxr {
+
+MultiSerialAgent::MultiSerialAgent(
+        uint8_t addr,
+        Middleware::Kind middleware_kind)
+    : Server<MultiSerialEndPoint>{middleware_kind}
+    , framing_io{}
+    , addr_{addr}
+    , buffer_{0}
+{
+    FD_ZERO(&read_fds);
+}
+
+void MultiSerialAgent::insert_serial(int serial_fd)
+{
+    utils::ExclusiveLock lk(framing_mtx);
+    FD_SET(serial_fd, &read_fds);
+    FramingIO aux_framing_io(addr_,
+        std::bind(&MultiSerialAgent::write_data, this, (uint8_t) serial_fd, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3),
+        std::bind(&MultiSerialAgent::read_data, this, (uint8_t) serial_fd, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, std::placeholders::_4));
+    
+    framing_io.insert(std::pair<int, FramingIO>(serial_fd, aux_framing_io));
+}
+
+bool MultiSerialAgent::remove_serial(int serial_fd)
+{
+    bool rv = false;
+    utils::ExclusiveLock lk(framing_mtx);
+
+    FD_CLR(serial_fd, &read_fds);
+    framing_io.erase(serial_fd);
+
+    if (0 == ::close(serial_fd))
+    {
+        UXR_AGENT_LOG_INFO(
+            UXR_DECORATE_GREEN("serial port closed"),
+            "fd: {}",
+            serial_fd);
+        rv = true;
+    }
+    else
+    {
+        UXR_AGENT_LOG_ERROR(
+            UXR_DECORATE_RED("close serial error"),
+            "fd: {}, errno: {}",
+            serial_fd, errno);
+    }
+
+    return rv;
+}
+
+bool MultiSerialAgent::recv_message(
+        InputPacket<MultiSerialEndPoint>& input_packet,
+        int timeout,
+        TransportRc& transport_rc)
+{
+    struct timeval timeout_ = { 1, 0 }; /* Seconds, Microseconds */
+    bool rv = false;
+
+    utils::SharedLock lk(framing_mtx);
+    fd_set fds = read_fds;
+    int ret = select(framing_io.rbegin()->first+1, &fds, NULL, NULL, &timeout_);
+    
+    if (0 < ret)
+    {
+        for (std::map<int, FramingIO>::iterator it = framing_io.begin(); it != framing_io.end(); it++)
+        {
+            if (FD_ISSET(it->first, &fds))
+            {
+                // Read it->first fd
+                uint8_t remote_addr;
+                ssize_t bytes_read = it->second.read_framed_msg(buffer_,sizeof (buffer_), remote_addr, timeout, transport_rc);
+
+                if (0 < bytes_read)
+                {
+                    input_packet.message.reset(new InputMessage(buffer_, static_cast<size_t>(bytes_read)));
+                    input_packet.source = MultiSerialEndPoint(it->first, remote_addr);
+                    rv = true;
+
+                    uint32_t raw_client_key;
+                    if (Server<MultiSerialEndPoint>::get_client_key(input_packet.source, raw_client_key))
+                    {
+                        UXR_AGENT_LOG_MESSAGE(
```

Comment:
Maybe here a reference to the serial descriptor.

Reply:
Updated 46a62d3:
```
[1621849003.217650] debug    | MultiSerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x4A705497, fd: 13, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 4A 02 00 00 80
[1621849003.218211] debug    | MultiSerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x50A5B13C, fd: 3, len: 20, data: 
0000: 81 80 49 02 07 01 08 00 02 53 00 05 45 02 00 00 00 00 08 40
[1621849003.218308] debug    | MultiSerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x50A5B13C, fd: 3, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 11 00 00 00 80
[1621849003.218478] debug    | MultiSerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x4A705497, fd: 13, len: 16, data: 
0000: 81 80 4A 02 07 01 08 00 02 54 00 05 46 02 00 00
```